### PR TITLE
Support projections in the manifest

### PR DESCRIPTION
**Pull request by dmvk**
This is the main building block for projection support. This PR only contains changes related to manifest and SST read path, it's still not exposed to the user.

### CODE REVIEW

File: schemas/sst.fbs [hunk 1]
```diff
@@ -1,13 +1,3 @@
-table CompactedSstId {
-    high: uint64;
-    low: uint64;
-}
-
-table CompactedSsTable {
-    id: CompactedSstId (required);
-    info: SsTableInfo (required);
-}
```
Comment (dmvk) [inline schemas/sst.fbs:9]: `sst.fbs` should only contain messages that are specific to the table format itself (= what's contained in the SST), this really belongs to the manifest

File: slatedb/src/bytes_range.rs [hunk 2]
```diff
@@ -23,105 +35,89 @@ pub(crate) fn is_prefix_increment(prefix: &[u8], b: &[u8]) -> bool {
...
+    where
+        T: RangeBounds<&'a [u8]>,
+    {
+        Self::new(
+            range.start_bound().map(|b| Bytes::copy_from_slice(b)),
+            range.end_bound().map(|b| Bytes::copy_from_slice(b)),
         )
     }
 
     #[cfg(test)]
```
Comment (dmvk) [inline slatedb/src/bytes_range.rs:82]: this will be used outside of tests as well in upcoming PRs

File: slatedb/src/sst_iter.rs [hunk 3]
```diff
@@ -45,22 +45,20 @@ impl Default for SstIteratorOptions {
 /// needed for [`crate::db::Db::scan`] since it returns the iterator, while [`SstView::Borrowed`]
 /// accommodates access by reference which is useful for [`crate::db::Db::get`].
 pub(crate) enum SstView<'a> {
-    Owned(SsTableHandle, BytesRange),
-    Borrowed(&'a SsTableHandle, (Bound<&'a [u8]>, Bound<&'a [u8]>)),
+    Owned(Box<SsTableHandle>, BytesRange),
```
Comment (dmvk) [inline slatedb/src/sst_iter.rs:48]: compiler started complaining that the size difference between Owned and Borrowed would be significant and suggested boxing the handle

File: schemas/manifest.fbs [hunk 4]
```diff
@@ -22,6 +22,20 @@ table ExternalDb {
     sst_ids: [CompactedSstId] (required);
 }
 
+table BytesRange {
+    // Start bound of the range. Empty bound means unbounded.
+    start_bound: [ubyte];
```
Comment (rodesai) [inline schemas/manifest.fbs]: I think this is technically fine because keys cannot be empty, so an empty bound value is not ambiguous. Still, did you consider a more general bound type with an enum for inclusive/exclusive/unbounded? e.g.

```
enum BoundType : byte { Unbounded = 1, Inclusive, Exclusive }

table BytesBound {
    bound: [ubyte];
    type: BoundType;
}

table BytesRange {
    start: BytesBound;
    end: BytesBound;
}
```
```

File: slatedb/src/bytes_range.rs [hunk 5]
```diff
@@ -23,105 +35,89 @@ pub(crate) fn is_prefix_increment(prefix: &[u8], b: &[u8]) -> bool {
...
 
-pub(crate) fn has_nonempty_intersection(
-    r1: (Bound<&[u8]>, Bound<&[u8]>),
-    r2: (Bound<&[u8]>, Bound<&[u8]>),
-) -> bool {
-    let start_bound = max_start_bound(r1.0, r2.0);
-    let end_bound = min_end_bound(r1.1, r2.1);
-    !is_empty(start_bound, end_bound)
+    #[cfg(test)]
+    pub(crate) fn is_empty(&self) -> bool {
```
Comment (rodesai) [inline slatedb/src/bytes_range.rs]: nit: name `empty` to keep consistent with other bool fns (which don't use `is_` prefix)

File: slatedb/src/bytes_range.rs [hunk 6]
```diff
@@ -23,105 +35,89 @@ pub(crate) fn is_prefix_increment(prefix: &[u8], b: &[u8]) -> bool {
...
+fn is_bound_non_empty(bound: &Bound<Bytes>) -> bool {
+    match bound {
+        Included(b) | Excluded(b) => !b.is_empty(),
+        Unbounded => true,
     }
 }
 
 impl BytesRange {
     pub(crate) fn new(start_bound: Bound<Bytes>, end_bound: Bound<Bytes>) -> Self {
+        assert!(
```
Comment (rodesai) [inline slatedb/src/bytes_range.rs:55]: why would you not be able to create a bound that includes/excludes the empty byte array?

File: slatedb/src/db_state.rs [hunk 7]
```diff
@@ -23,50 +23,99 @@ use SsTableId::{Compacted, Wal};
...
+    #[cfg(test)]
+    pub(crate) fn with_visible_range(&self, visible_range: BytesRange) -> Self {
+        Self::new_compacted(self.id, self.info.clone(), Some(visible_range))
     }
 
     // Compacted (non-WAL) SSTs are never empty. They are created by compaction or
-    // memtable flushes, which should never produce empty SSTs.
-    pub(crate) fn compacted_first_key(&self) -> &Bytes {
+    // memtable flushes, which should never produce empty SSTs. This method returns
+    // the start key after applying projections.
```
Comment (rodesai) [inline slatedb/src/db_state.rs:104]: not for this PR but we should probably have a separate type for non-empty SSTs (basically all compacted SSTs and WAL SSTs other than the fencing ones)

File: slatedb/src/sst_iter.rs [hunk 8]
```diff
@@ -168,7 +164,13 @@ impl<'a> SstIterator<'a> {
         table_store: Arc<TableStore>,
         options: SstIteratorOptions,
     ) -> Result<Self, SlateDBError> {
-        let view = SstView::Owned(table, BytesRange::from(range));
+        let mut view_range = BytesRange::from(range);
+        if let Some(visible_range) = &table.visible_range {
+            view_range = view_range
+                .intersect(visible_range)
+                .expect("Provided range is outside of table's visible range");
```
Comment (rodesai) [inline slatedb/src/sst_iter.rs]: why are we panic'ing in this case? I think it's expected to hit this case if the client tries to do a scan outside the db's projected range - in which case we should just return an empty iterator.

File: slatedb/src/sst_iter.rs [hunk 9]
```diff
@@ -168,7 +164,13 @@ impl<'a> SstIterator<'a> {
         table_store: Arc<TableStore>,
         options: SstIteratorOptions,
     ) -> Result<Self, SlateDBError> {
-        let view = SstView::Owned(table, BytesRange::from(range));
+        let mut view_range = BytesRange::from(range);
+        if let Some(visible_range) = &table.visible_range {
```
Comment (rodesai) [inline slatedb/src/sst_iter.rs]: nit: seems better to make this a fn of `SsTableHandle`

File: slatedb/src/bytes_range.rs [hunk 10]
```diff
@@ -1,15 +1,27 @@
...
 /// the cumbersome work associated with the generic trait RangeBounds<Bytes>
-#[derive(Debug, Clone, PartialEq)]
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
 pub(crate) struct BytesRange {
-    start_bound: Bound<Bytes>,
-    end_bound: Bound<Bytes>,
+    inner: ComparableRange<Bytes>,
+}
+
+impl Serialize for BytesRange {
```
Comment (rodesai) [inline slatedb/src/bytes_range.rs:15]: where are we serializing this type?

File: slatedb/src/db_state.rs (hunk 7 above)
Comment (dmvk) [reply slatedb/src/db_state.rs:104]: 💯, I've thought about the same thing, until I've found out how fencing works this seemed super confusing

File: slatedb/src/bytes_range.rs (hunk 10 above)
Comment (dmvk) [reply slatedb/src/bytes_range.rs:15]: In admin CLI - https://github.com/slatedb/slatedb/blob/main/slatedb/src/admin.rs#L43

The type is now part of the manifest and we need to be able to convert manifest to json for debug purpose.

Comment (dmvk) [reply slatedb/src/bytes_range.rs:15]: The trait is implemented here (instead of derived) to make the output more readable.

File: slatedb/src/db_state.rs [hunk 11]
```diff
@@ -23,50 +23,99 @@ use SsTableId::{Compacted, Wal};
 pub(crate) struct SsTableHandle {
     pub id: SsTableId,
     pub info: SsTableInfo,
+    pub visible_range: Option<BytesRange>,
```
Comment (criccomini) [inline slatedb/src/db_state.rs:29]: Can you add some rustdocs here. I know I'm going to forget the meaning. Plus... AI needs help.

File: slatedb/src/db_state.rs [hunk 12]
```diff
@@ -23,50 +23,99 @@ use SsTableId::{Compacted, Wal};
...
+        id: SsTableId,
+        info: SsTableInfo,
+        visible_range: Option<BytesRange>,
+    ) -> Self {
+        let mut effective_range = match info.first_key.clone() {
+            Some(physical_first_key) => {
+                BytesRange::new(Included(physical_first_key.clone()), Unbounded)
+            }
+            None => {
+                unreachable!("SST always has a first key.")
```
Comment (criccomini) [inline slatedb/src/db_state.rs:77]: 🤔 Is this true? I wonder what the compactor does if the only `RowEntry`s in the SST are deletes....

File: slatedb/src/size_tiered_compaction.rs [hunk 13]
```diff
@@ -633,10 +633,7 @@ mod tests {
             filter_len: 0,
             compression_codec: None,
         };
-        SsTableHandle {
-            id: SsTableId::Compacted(crate::utils::ulid()),
-            info,
-        }
+        SsTableHandle::new(SsTableId::Compacted(ulid::Ulid::new()), info)
```
Comment (criccomini) [inline slatedb/src/size_tiered_compaction.rs]: Please use `crate::utlis::ulid()`. I'll update `clippy.toml` to catch this.

**Review by criccomini — COMMENTED** [review]
LGTM. Couple of random comments, but this makes sense.

File: schemas/manifest.fbs [hunk 14]
```diff
@@ -74,6 +88,17 @@ table ManifestV1 {
...
 
+table CompactedSstId {
+    high: uint64;
+    low: uint64;
+}
+
+table CompactedSsTable {
+    id: CompactedSstId (required);
+    info: SsTableInfo (required);
+    visible_range: BytesRange;
```
Comment (vigneshc) [inline schemas/manifest.fbs:103]: Is `visible_range` always required? Shouldn't it be an option?

File: schemas/manifest.fbs [hunk 15]
```diff
@@ -74,6 +88,17 @@ table ManifestV1 {
     wal_object_store_uri: string;
 }
 
+table CompactedSstId {
+    high: uint64;
+    low: uint64;
+}
+
+table CompactedSsTable {
```
Comment (vigneshc) [inline schemas/manifest.fbs:100]: Would this cause any issue for SSTs created in older SlateDB versions?

**Comment by vigneshc** [comment]
I assume this PR is part of the DB split and DB merge work.
This lays the foundation for split. I think this is Chris meant in his comment in #577 . 

How would merge be implemented? It is out of scope for this PR, but curious about how you were thinking of solving it.

File: slatedb/src/db_state.rs (hunk 11 above)
Comment (vigneshc) [reply slatedb/src/db_state.rs:29]: What is the difference between `visible_range` and `effective_range`?
Isn't `visible_range`, the `effective_range`? Or does it account for the actual `first_key`?

File: slatedb/src/sst_iter.rs (hunk 8 above)
Comment (dmvk) [reply slatedb/src/sst_iter.rs]: This one was tricky, I did few experiments and the cleanest solution feels to be returning None from this method if iterator falls outside of visible range.

https://github.com/slatedb/slatedb/pull/593/commits/266205a090a5e178c5f79364f23de974ea8d7247

WDYT?

File: slatedb/src/db_state.rs (hunk 11 above)
Comment (dmvk) [reply slatedb/src/db_state.rs:29]: effective range is intersection of visible_range (projection) and physical_range (`first_key..`)

File: schemas/manifest.fbs (hunk 14 above)
Comment (dmvk) [reply schemas/manifest.fbs:103]: it's not required meaning it's optional. This is a flatbuffer definition.

File: schemas/manifest.fbs (hunk 15 above)
Comment (dmvk) [reply schemas/manifest.fbs:100]: I'm not sure what exactly you're referring to. There are basically two changes.

First is moving CST into different file, which doesn't even change generated code, so that's harmless.

Second thing we're doing is adding a new optional field, that should also be safe from compatibility PoV.

Grok:

> Adding a new optional field to a FlatBuffer schema is not a breaking change for forward compatibility, but it can be a breaking change for backward compatibility, depending on how the schema is used. Here's a concise explanation:
>
> Forward Compatibility: Existing code reading an updated schema with a new optional field will ignore the new field if it’s not recognized. FlatBuffers are designed to skip unknown fields, so older consumers can still process the data without issues.
Backward Compatibility: New code reading old data (serialized with the older schema) will see the new optional field as absent (null or default value). If the new code assumes the optional field is always present or mishandles its absence, it could lead to errors. Proper handling of optional fields (e.g., checking for null or defaults) ensures compatibility.
Best Practices:
>
> Always mark new fields as optional to avoid breaking older consumers.
Ensure new code gracefully handles cases where the optional field is missing.
Avoid making the new field required, as that would be a breaking change for both forward and backward compatibility.
If you follow these practices, adding an optional field is generally safe and non-breaking for forward compatibility, but you must carefully handle backward compatibility in your code.

File: slatedb/src/db_state.rs (hunk 12 above)
Comment (dmvk) [reply slatedb/src/db_state.rs:77]: As far as I can tell for compacted SSTs yes. Only case where SST might not have it is WAL Fence.

https://github.com/slatedb/slatedb/pull/593#discussion_r2131555568

Comment (dmvk) [reply slatedb/src/db_state.rs:77]: here is a test case; what's little weird is that we'll end up with "empty sorted run" entry, but that's most likely harmless

Comment (dmvk) [reply slatedb/src/db_state.rs:77]: https://github.com/dmvk/slatedb/commit/bfd4182aecbf6bd7f693a488e60b8dce33d9550e

File: slatedb/src/db_state.rs (hunk 11 above)
Comment (dmvk) [reply slatedb/src/db_state.rs:29]: added comments for both

File: slatedb/src/bytes_range.rs (hunk 6 above)
Comment (dmvk) [reply slatedb/src/bytes_range.rs:55]: Because it's super tricky to reason about and it introduces bunch of test cases we have to test for, while not providing any value to the user. We've already talked about that with @criccomini on discord, I'll try to dig up the thread.

Let's think about following cases:

```
`..a`  vs. `""..a`
`a..` vs. `a..""`
`..` vs. `""..""` // I don't even know how this should behave
```

The main problem is that ordering of empty byte is not really defined + it is not a valid key. I assume that any time someone actually uses it, they really mean unbounded, which we have separate representation for.

Overall I think placing such invariant into the code will make our live way easier down the road (this simple change can already eliminate bunch of test suites covering for all sorts of edge cases).

WDYT? Is there any use case I'm missing?

**Comment by dmvk** [comment]
@vigneshc 

> How would merge be implemented? It is out of scope for this PR, but curious about how you were thinking of solving it.

You can get a sneak peak at https://github.com/slatedb/slatedb/pull/493. Merging will more or less be a special case of database clone function.

File: schemas/manifest.fbs (hunk 4 above)
Comment (dmvk) [reply schemas/manifest.fbs]: No, I haven't really thought about this much. I actually think I've copied the structure from some other project, but I was some time ago so I'm not able to find it anymore.

I'm fine with your approach, here is how it would look like, lmk what you think: https://github.com/dmvk/slatedb/commit/e4fdf0bdb4fb351eaa87aff3e358bc1fdff87ca2

File: slatedb/src/bytes_range.rs (hunk 6 above)
Comment (dmvk) [reply slatedb/src/bytes_range.rs:55]: One test that occasionally fails is related to this (proptest that generates such range), let's first close on this discussion before I fix it.

**Comment by dmvk** [comment]
This is ready for another pass.

File: slatedb/src/bytes_range.rs (hunk 6 above)
Comment (dmvk) [reply slatedb/src/bytes_range.rs:55]: https://discord.com/channels/1232385660460204122/1341139884148260864/1341141579942858753

This is the discussion I was referring to

File: slatedb/src/db_state.rs (hunk 12 above)
Comment (criccomini) [reply slatedb/src/db_state.rs:77]: Interesting.. good to know! I guess it makes sense that it wouldn't write the SST if it's all deletes. As you said, the empty SR is odd but seems relatively harmless. I wonder if we have logic to drop them, or if they keep accruing in the manifest. 🤔 

Anyway, thanks for checking!

File: schemas/manifest.fbs (hunk 4 above)
Comment (criccomini) [reply schemas/manifest.fbs]: LGTM. I like having all three states modeled (and unknown, too!)

File: schemas/manifest.fbs (hunk 15 above)
Comment (criccomini) [reply schemas/manifest.fbs:100]: Yep, FB optional field additions are safe _as long as they're at the end of the schema_.

https://flatbuffers.dev/evolution/#addition

This change should be safe.

**Review by criccomini — COMMENTED** [review]
This LGTM, but I'm going to leave @rodesai to do the final ✅ just so we have two 👀 on it.

File: schemas/manifest.fbs (hunk 4 above)
Comment (rodesai) [reply schemas/manifest.fbs]: Thanks! This looks great

**Comment by criccomini** [comment]
(Had to close/open because GH was having some issue with head when merging)

File: slatedb/src/bytes_range.rs (hunk 6 above)
Comment (criccomini) [reply slatedb/src/bytes_range.rs:55]: This came up again here: https://github.com/slatedb/slatedb/pull/1461
